### PR TITLE
[tpu-inference] Add batched block-diffusion program

### PR DESCRIPTION
**Pull request by ethannnnnn**
> [!WARNING]
> These denoising modules are experimental and live under
> `tpu_inference.runner.experimental.diffusion`. They are not selected by the
> existing runner in this PR.

## Motivation

The denoising algorithm should be testable independently from serving state, scheduler integration, and a concrete model forward implementation.

## Scope

- Add low-confidence token commitment with guaranteed forward progress.
- Add a batched JAX denoising program for active and padded request rows.
- Support configured logit alignment, sub-block processing, step limits, KV updates, and next-block anchors.

## Design

The jitted program accepts model-forward and token-commit callables. Each sub-block repeatedly commits positions above the confidence threshold, forces one eligible position when necessary, and performs a final forward pass to commit KV state and derive the next anchor.

## Compatibility

Autoregressive behavior is unchanged. These are new diffusion modules and are not selected or invoked by the existing runner in this change.

## Extensibility

Callable algorithm and forward boundaries keep denoising independent of the model implementation. Policy enums from the foundation change control alignment and anchor behavior without branching on model names.

## Tests

- Relevant CPU test selection at this stack point: 24 passed.
- Stack checks passed: `isort`, `yapf`, `ruff`, `py_compile`, and `git diff --check`.

## Known limitations

Only the low-confidence commit algorithm is implemented. Serving and scheduler integration are intentionally left to later PRs in the stack.

## Stack

- Depends on the previous upstream PR in this stack: https://github.com/vllm-project/tpu-inference/pull/3243. Until that PR merges, GitHub's `main`-based comparison includes the preceding stack commits.
- [Public design document](https://docs.google.com/document/d/1d3qtGrvdcJ0ptluE0EwAaGj4AQyOlIYkYvKo9xPguX8/edit)

### CODE REVIEW

File: tpu_inference/runner/diffusion/program.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+def _align_logits(
+    logits: jax.Array,
+    alignment: LogitAlignment,
+) -> jax.Array:
+    if alignment is LogitAlignment.SAME_POSITION:
+        return logits
+    if alignment is LogitAlignment.SHIFTED:
+        indices = jnp.maximum(
+            jnp.arange(logits.shape[1], dtype=jnp.int32) - 1, 0)
```
Comment (syhuang22) [inline tpu_inference/runner/diffusion/program.py:49]: The `jnp.maximum(arange - 1, 0)` clamp makes position 0 consume its ownlogits, which under SHIFTED semantics predict position 1. If position 0 is ever masked (eligible), it will commit the wrong token with no error.

File: tpu_inference/runner/diffusion/program.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        return canvas, mask, denoise_steps, kv
+
+    canvas, mask, denoise_steps, kv_caches = jax.lax.fori_loop(
+        0,
+        num_sub_blocks,
+        denoise_sub_block,
+        (canvas, mask, denoise_steps, kv_caches),
+    )
+
+    final_logits, kv_caches = forward_fn(model_state, canvas, positions,
```
Comment (syhuang22) [inline tpu_inference/runner/diffusion/program.py:173]: Implementation looks correct (final forward runs on the fully committed canvas). But the only test just checks final KV == recomputation. Suggest to add a propagation test: make the last position resolve via forced-commit only (no denoise forward ever saw the final token), then assert the next block's output changes vs. skipping the final forward.

File: tpu_inference/runner/diffusion/program.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    if sub_block_size < 1 or initial_canvas.shape[1] % sub_block_size:
+        raise ValueError("sub_block_size must divide the model block size")
+
+    batch_size, block_size = initial_canvas.shape
+    del batch_size
+    active_rows = jnp.asarray(active_rows, dtype=bool)
+    canvas = initial_canvas.astype(jnp.int32)
+    mask = jnp.asarray(initial_mask, dtype=bool) & active_rows[:, None]
+    denoise_steps = jnp.zeros((canvas.shape[0], ), dtype=jnp.int32)
+    steps_per_sub_block = (sub_block_size
```
Comment (syhuang22) [inline tpu_inference/runner/diffusion/program.py:111]: `max_denoise_steps` is per-*sub-block*, not per-block (4 sub-blocks x 8 =
   up to 32 forwards). Please document the unit or rename.

File: tpu_inference/runner/diffusion/algorithm.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    if logits.ndim != 3:
+        raise ValueError("logits must have shape [batch, length, vocab]")
+    if eligible_mask.shape != logits.shape[:2]:
+        raise ValueError("eligible_mask must match logits [batch, length]")
+
+    active_rows = jnp.asarray(active_rows, dtype=bool)
+    confidence_threshold = jnp.asarray(confidence_threshold, dtype=jnp.float32)
+    temperature = jnp.asarray(temperature, dtype=jnp.float32)
+    safe_temperature = jnp.where(temperature > 0.0, temperature, 1.0)
+    scaled_logits = logits.astype(jnp.float32) / safe_temperature[:, None,
```
Comment (syhuang22) [inline tpu_inference/runner/diffusion/algorithm.py:77]: temperature only rescales confidence — argmax is scale-invariant and there's no RNG, so it never changes token selection. Please state this in the docstring.

File: tpu_inference/runner/diffusion/program.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            canvas = jnp.where(committed, token_ids, canvas)
+            mask &= ~committed
+            steps += row_has_work.astype(jnp.int32)
+            return (canvas, mask, steps, kv, remaining, token_ids,
+                    iteration + 1)
+
+        state = jax.lax.while_loop(has_work, denoise_step, state)
+        canvas, mask, denoise_steps, kv, remaining, last_tokens, _ = state
+
+        canvas = jnp.where(remaining, last_tokens, canvas)
```
Comment (syhuang22) [inline tpu_inference/runner/diffusion/program.py:162]: cap-exit tail bulk-commits all remaining positions from the last argmax — a quality cliff, and dead code in every current test. Add a test forcing it (`max_denoise_steps < sub_block_size`).

**Comment by ethannnnnn** [comment]
Addressed the current review feedback and rebased the complete six-PR stack onto the latest `main`.

Key changes in this update:
- use the model-neutral `seeded_shifted` adapter and generic `block_size` metadata;
- prevent the MASK token from being selected as a committed token or next-block anchor;
- reject unsupported shifted/all-masked semantics and fail if an active shifted canvas masks its seed;
- document that `max_denoise_steps` is per sub-block and that temperature changes confidence, not greedy token selection;
- test inactive-row KV behavior, forced progress, and next-block propagation from the final KV refresh;
- propagate the API update through the serving strategy and platform tests.

Validation: 41 focused CPU tests passed; ruff, isort, yapf, compile, whitespace, and public-text checks passed.

Design: https://docs.google.com/document/d/1d3qtGrvdcJ0ptluE0EwAaGj4AQyOlIYkYvKo9xPguX8/edit

Vanilla public reproduction guide: https://docs.google.com/document/d/1rjRSopy888ZqLnGh_KUseW8h1_ZqPGEuSee5MM2Q9ac/edit

@syhuang22, could you please take another look?
